x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/textstructure/action/TestGrokPatternAction.java

```diff
@@ -0,0 +1,194 @@
+            super(in);
+            grokPattern = in.readString();
+            text = in.readStringCollectionAsList();
+        }
+
+        public static Request parseRequest(XContentParser parser) throws IOException {
+            return PARSER.parse(parser, null).build();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(grokPattern);
+            out.writeStringCollection(text);
+        }
+
+        public String getGrokPattern() {
+            return grokPattern;
+        }
+
+        public List<String> getText() {
+            return text;
+        }
+
+        @Override
+        public ActionRequestValidationException validate() {
+            ActionRequestValidationException validationException = null;
+            if (grokPattern == null) {
+                validationException = addValidationError("[" + GROK_PATTERN.getPreferredName() + "] missing", validationException);
+            }
+            if (text == null) {
+                validationException = addValidationError("[" + TEXT.getPreferredName() + "] missing", validationException);
+            }
+            return validationException;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Request request = (Request) o;
+            return Objects.equals(grokPattern, request.grokPattern) && Objects.equals(text, request.text);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(grokPattern, text);
+        }
+
+        @Override
+        public String toString() {
+            return "Request{" + "grokPattern='" + grokPattern + '\'' + ", text=" + text + '}';
+        }
+    }
+
+    public static class Response extends ActionResponse implements ToXContentObject, Writeable {
+
+        private final List<Map<String, Object>> ranges;
+
+        public Response(List<Map<String, Object>> ranges) {
+            this.ranges = ranges;
+        }
+
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            ranges = in.readCollectionAsList(StreamInput::readGenericMap);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.startArray("matches");
+            for (Map<String, Object> ranges : ranges) {
+                builder.startObject();
+                builder.field("matched", ranges != null);
+                if (ranges != null) {
+                    builder.startObject("fields");
+                    for (Map.Entry<String, Object> rangeOrList : ranges.entrySet()) {
+                        if (rangeOrList.getValue() instanceof GrokCaptureExtracter.Range) {
```

Comment:
There's still an open discussion about whether it's better that this case be returned as an array with one element rather than not an array at all.

Reply:
I think it’s best that we always return an array. In some languages that will avoid the developer using our client having to have two code branches, one that deals with an array and one that deals with a single object. Callers who know for sure they’ll only ever get a single element array can just get the first element of the array without checking its length to simplify their code.

Reply:
Done!